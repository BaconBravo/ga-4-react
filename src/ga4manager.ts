declare global {
  interface Window {
    gtag: Function;
  }
}

/**
 * @interface
 */
export interface GA4ReactInterface extends GA4ReactResolveInterface {
  initialize(): Promise<any>;
}
export interface GA4ReactResolveInterface {
  pageview(path: string): void;
  gtag(...args: any): any;
}

/**
 * @desc class required to manage google analitycs 4
 * @class GA4React
 *  */
export class GA4React implements GA4ReactInterface {
  constructor(
    private gaCode: string,
    private config?: object,
    private additionalGaCode?: Array<string>
  ) {
    this.config = config || {};
    this.gaCode = gaCode;
  }
  public initialize(): Promise<GA4ReactResolveInterface> {
    return new Promise((resolve, reject) => {
      const head: HTMLHeadElement = document.getElementsByTagName('head')[0];
      const scriptAsync: HTMLScriptElement = document.createElement('script');
      scriptAsync.setAttribute('async', '');
      scriptAsync.setAttribute(
        'src',
        `https://www.googletagmanager.com/gtag/js?id=${this.gaCode}`
      );
      scriptAsync.onload = () => {
        const scriptSync: HTMLScriptElement = document.createElement('script');

        let scriptHTML: string = `
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', '${this.gaCode}', ${JSON.stringify(this.config)});`;

        if (this.additionalGaCode) {
          this.additionalGaCode.forEach((code: string) => {
            scriptHTML += `gtag('config', '${code}');`;
          });
        }

        scriptSync.innerHTML = scriptHTML;

        head.appendChild(scriptSync);
        resolve({ pageview: this.pageview, gtag: this.gtag });
      };

      scriptAsync.onerror = (err: any) => {
        reject(err);
      };

      head.appendChild(scriptAsync);
    });
  }

  /**
   * @desc set new page or send pageview event
   * @param path
   */
  public pageview(path: string): void {
    window.gtag('event', 'page_view', {
      page_path: path,
      page_location: window.location,
      page_title: document.title,
    });
  }

  /**
   * @desc direct access to gtag
   * @param args
   */
  public gtag(...args: any): any {
    return window.gtag(args);
  }
}

export default GA4React;
