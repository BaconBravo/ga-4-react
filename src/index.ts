declare global {
  interface Window {
    gtag: Function;
  }
}

/**
 * @interface
 */
export interface GA4ReactInterface {
  initialize(): Promise<any>;
  pageview(path: string): void;
  gtag(...args: any): any;
}

/**
 * @desc class required to manage google analitycs 4
 * @class GA4React
 *  */
export class GA4React implements GA4ReactInterface {
  constructor(
    private gaCode: string,
    private config?: object,
    private additionalGaCode?: Array<string>
  ) {
    this.config = config || {};
    this.gaCode = gaCode;
  }
  initialize(): Promise<GA4ReactInterface> {
    return new Promise((resolve, reject) => {
      const head: HTMLHeadElement = document.getElementsByTagName('head')[0];
      const scriptAsync: HTMLScriptElement = document.createElement('script');
      scriptAsync.setAttribute('async', '');
      scriptAsync.setAttribute(
        'src',
        `https://www.googletagmanager.com/gtag/js?id=${this.gaCode}`
      );
      scriptAsync.onload = () => {
        const scriptSync: HTMLScriptElement = document.createElement('script');

        let scriptHTML: string = `
          window.dataLayer = window.dataLayer || [];
          function gtag(){dataLayer.push(arguments);}
          gtag('js', new Date());
          gtag('config', '${this.gaCode}', ${JSON.stringify(this.config)});`;

        if (this.additionalGaCode) {
          this.additionalGaCode.forEach((code: string) => {
            scriptHTML += `gtag('config', '${code}');`;
          });
        }

        scriptSync.innerHTML = scriptHTML;

        head.appendChild(scriptSync);
        resolve(this);
      };

      scriptAsync.onerror = (err: any) => {
        reject(err);
      };

      head.appendChild(scriptAsync);
    });
  }

  /**
   * @desc set new page or send pageview event
   * @param path
   */
  pageview(path: string): void {
    window.gtag('event', 'page_view', {
      page_path: path,
      page_location: window.location,
      page_title: document.title,
    });
  }

  /**
   * @desc direct access to gtag
   * @param args
   */
  gtag(...args: any): any {
    return window.gtag(args);
  }
}

export default GA4React;
