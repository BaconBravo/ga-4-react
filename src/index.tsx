import React, { useEffect, useState } from 'react';
import GA4React from './ga4manager';

export interface IGAReactConfig {
  send_page_view: boolean;
  groups: string;
}

export interface IGA4R {
  code: string;
  config?: IGAReactConfig;
  additionalCode?: Array<string>;
  children?: any;
}

export const GA4R: React.FC<IGA4R> = ({
  code,
  config,
  additionalCode,
  children,
}) => {
  const [components, setComponents] = useState<any>(children);

  useEffect(() => {
    const ga4manager = new GA4React(`${code}`, config, additionalCode);
    ga4manager.initialize().then(
      ga4 => {
        setComponents(
          React.Children.map(children, (child, index) => {
            return React.cloneElement(child, {
              ga4: ga4,
              index,
            });
          })
        );
      },
      err => {
        console.error(err);
      }
    );
  }, []);

  return <>{components}</>;
};

export default GA4React;
