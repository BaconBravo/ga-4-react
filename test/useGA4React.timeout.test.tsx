import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { useGA4React } from '../src/index';

Object.defineProperty(global.document, 'readyState', {
  get() {
    return 'interactive';
  },
});

beforeAll(() => {
  global.document.head.innerHTML = '';
});

const Example = () => {
  const ga4React = useGA4React('GA-CODE', {}, [], 10);
  return <>{JSON.stringify(ga4React)}</>;
};

describe('GA4R hook', () => {
  it('Rendering with very short timeout', async done => {
    const { container } = render(<Example />);

    setTimeout(() => {
      global.document.dispatchEvent(new Event('readystatechange'));
    }, 100);

    setTimeout(() => {
      const LoadEvent = document.createEvent('HTMLEvents');
      LoadEvent.initEvent('load', true, true);
      const target = global.document.head.querySelector('script');
      if (target) {
        target.dispatchEvent(LoadEvent);
      }
    }, 1000);

    setTimeout(() => {
      expect(container.innerHTML).toMatchSnapshot();
      expect(global.document.head).toMatchSnapshot();
      done();
    }, 2000);
  });
});
