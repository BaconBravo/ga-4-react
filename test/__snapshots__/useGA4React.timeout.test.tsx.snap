// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GA4R hook Rendering with very short timeout 1`] = `""`;

exports[`GA4R hook Rendering with very short timeout 2`] = `
<head>
  <script
    async=""
    id="ga4ReactScriptAsync"
    src="https://www.googletagmanager.com/gtag/js?id=GA-CODE"
  />
  <script
    id="ga4ReactScriptSync"
  >
    window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);};
        gtag('js', new Date());
        gtag('config', 'GA-CODE', {});
  </script>
</head>
`;
